cmd/s3-zip-handlers.go

```diff
@@ -0,0 +1,496 @@
+		ModTime: zipObjInfo.ModTime,
+	}
+
+	var rc io.ReadCloser
+
+	if file.UncompressedSize64 > 0 {
+		rs := &HTTPRangeSpec{Start: file.Offset, End: file.Offset + int64(file.UncompressedSize64) - 1}
+		gr, err := objectAPI.GetObjectNInfo(ctx, bucket, zipPath, rs, nil, readLock, ObjectOptions{})
+		if err != nil {
+			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+			return
+		}
+		defer gr.Close()
+		rc, err = file.Open(gr)
+		if err != nil {
+			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+			return
+		}
+	} else {
+		rc = stdioutil.NopCloser(bytes.NewReader([]byte{}))
+	}
+
+	defer rc.Close()
+
+	if err = setObjectHeaders(w, fileObjInfo, nil, opts); err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		return
+	}
+
+	setHeadGetRespHeaders(w, r.URL.Query())
+
+	httpWriter := ioutil.WriteOnClose(w)
+
+	// Write object content to response body
+	if _, err = io.Copy(httpWriter, rc); err != nil {
+		if !httpWriter.HasWritten() {
+			// write error response only if no data or headers has been written to client yet
+			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+			return
+		}
+		if !xnet.IsNetworkOrHostDown(err, true) { // do not need to log disconnected clients
+			logger.LogIf(ctx, fmt.Errorf("Unable to write all the data to client %w", err))
+		}
+		return
+	}
+
+	if err = httpWriter.Close(); err != nil {
+		if !httpWriter.HasWritten() { // write error response only if no data or headers has been written to client yet
+			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+			return
+		}
+		if !xnet.IsNetworkOrHostDown(err, true) { // do not need to log disconnected clients
+			logger.LogIf(ctx, fmt.Errorf("Unable to write all the data to client %w", err))
+		}
+		return
+	}
+}
+
+// listObjectsV2InArchive generates S3 listing result ListObjectsV2Info from zip file, all parameters are already validated by the caller.
+func listObjectsV2InArchive(ctx context.Context, objectAPI ObjectLayer, bucket, prefix, token, delimiter string, maxKeys int, fetchOwner bool, startAfter string) (ListObjectsV2Info, error) {
+	zipPath, _, err := splitZipExtensionPath(prefix)
+	if err != nil {
+		// Return empty listing
+		return ListObjectsV2Info{}, nil
+	}
+
+	zipObjInfo, err := objectAPI.GetObjectInfo(ctx, bucket, zipPath, ObjectOptions{})
+	if err != nil {
+		// Return empty listing
+		return ListObjectsV2Info{}, nil
+	}
+
+	var zipInfo []byte
+
+	if z, ok := zipObjInfo.UserDefined[archiveInfoMetadataKey]; ok {
+		zipInfo = []byte(z)
+	} else {
+		zipInfo, err = updateObjectMetadataWithZipInfo(ctx, objectAPI, bucket, zipPath)
+	}
+
+	if err != nil {
+		return ListObjectsV2Info{}, err
+	}
+
+	files, err := zipindex.DeserializeFiles(zipInfo)
+	if err != nil {
+		return ListObjectsV2Info{}, err
+	}
+
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name < files[j].Name
+	})
```

Comment:
how expensive is it to deserialize each time for listing? @vadmeste @klauspost - we also perhaps need to indicate that memory might be pretty big and perhaps do not allow listing on really big zip archives?